src/logic/basic.lean

```diff
@@ -577,6 +577,47 @@ iff_true_intro $ λ h, hn.elim h
 @[simp] theorem exists_prop_of_false {p : Prop} {q : p → Prop} : ¬ p → ¬ (∃ h' : p, q h') :=
 mt Exists.fst
 
+lemma exists_unique.exists {α : Sort*} {p : α → Prop} (h : ∃! x, p x) : ∃ x, p x :=
+exists.elim h (λ x hx, ⟨x, and.left hx⟩)
+
+lemma exists_unique.unique {α : Sort*} {p : α → Prop} (h : ∃! x, p x)
+  {y₁ y₂ : α} (py₁ : p y₁) (py₂ : p y₂) : y₁ = y₂ :=
+unique_of_exists_unique h py₁ py₂
+
+@[simp] lemma exists_unique_iff_exists {α : Sort*} [subsingleton α] {p : α → Prop} :
+  (∃! x, p x) ↔ ∃ x, p x :=
+⟨λ h, h.exists, Exists.imp $ λ x hx, ⟨hx, λ y _, subsingleton.elim y x⟩⟩
+
+lemma exists_unique.elim2 {α : Sort*} {p : α → Sort*} [∀ x, subsingleton (p x)]
+  {q : α → Prop} {b : Prop} (h₂ : ∃! x (h : p x), q x)
```

Comment:
This is a very funny usage of `exists_unique`. I know that nested `exists_unique` doesn't usually mean what it's supposed to mean, but I guess this is okay.
 
I think `q` should maybe have type `Π a : α, p a → Prop` to make this slightly more general.

Reply:
Same comment for similar lemmas

Reply:
Since `p x` is a subsingleton, `∃! x (h : p x), q x` is the same as `∃! x, p x ∧ q x`. This is not `defeq` but `∃! x ∈ s, p x` is more readable than `∃ x ∈ s, (p x ∧ ∀ y ∈ s, p y → y = x)`.

Reply:
Done.